### PR TITLE
Add transformer simulation features

### PR DESCRIPTION
**Pull request by ariefrahmansyah**
**What this PR does / why we need it**:

Enable users to test their standard transformer configuration without actual deployment.
The goal of this features is to shorten the development process.

- Add simulate API
- Add executor package as an "interface" to the transformer
- Revamp deployment UI to cater simulation components
- Revamp transformer tools

**Does this PR introduce a user-facing change?**:

```release-note
* Add Simulation components on deployment page
* Modify transformer tools page
```

Deployment Page:

image

Transformation Tools (merlin/-/tools/transformer) page:
image

**Checklist**

- [ ] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/pkg/transformer/pipeline/operation.go [hunk 1]
```diff
@@ -1,7 +1,114 @@
...
+}
+
+// GetOperationTracingDetail retrieve all the tracing detail
+// it will flatten the result if the operation specs is in slice form
+func (ot *OperationTracing) GetOperationTracingDetail() ([]types.TracingDetail, error) {
+	refVal := reflect.ValueOf(ot.Specs)
+	if refVal.Kind() == reflect.Slice {
+		numOfSpecs := refVal.Len()
+		if len(ot.Input) != numOfSpecs {
+			return nil, fmt.Errorf("number of inputs is not match with number of specs")
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/operation.go:84]: Can you update the error message? Ideally it should mention which operation, expected value, and the actual value

**Comment by pradithya** [comment]
The API part is looking good. I will test the UI.

File: api/config/config.go [hunk 2]
```diff
@@ -205,8 +205,45 @@ type StandardTransformerConfig struct {
...
 }
 
+// SimulationFeastConfig feast config that aimed to be used only for simulation of standard transformer
+type SimulationFeastConfig struct {
+	FeastRedisURL    string `envconfig:"SIMULATION_FEAST_REDIS_URL" required:"true"`
+	FeastBigtableURL string `envconfig:"SIMULATION_FEAST_BIGTABLE_URL" required:"true"`
+}
+
+// ToFeastStorageConfigsForSimulation convert standard transformer config to feast storage config that will be used for transformer simulation
+// the difference with ToFeastStorageConfigs, this method will overwrite serving url by using serving url that can be access outside of model k8s cluster and serving type is grpc (direct storage is not supported)
```
Comment (karzuo) [inline api/config/config.go:218]: break down to multiple linkes <120 chars

File: api/config/config_test.go [hunk 3]
```diff
@@ -52,9 +52,147 @@ func TestFeastServingURLs_URLs(t *testing.T) {
 	}
 }
 
-func TestStandardTransfomer_ToFeastStorageConfigs(t *testing.T) {
+func TestStandardTransformerConfig_ToFeastStorageConfigsForSimulation(t *testing.T) {
 	redisCfg := `{"is_redis_cluster": true,"serving_url":"online-storage.merlin.dev","redis_addresses":["10.1.1.10", "10.1.1.11"],"pool_size": 4,"max_retries": 1,"dial_timeout": "10s"}`
```
Comment (karzuo) [inline api/config/config_test.go:56]: breakdown this line too and the next

File: api/config/config_test.go [hunk 4]
```diff
@@ -52,9 +52,147 @@ func TestFeastServingURLs_URLs(t *testing.T) {
...
+			err := envconfig.Process("", &cfg)
+			require.NoError(t, err)
+			got := cfg.ToFeastStorageConfigsForSimulation()
+			assert.Equal(t, tC.feastStorageCfg, got)
+		})
+	}
+}
+
+func TestStandardTransformerConfig_ToFeastStorageConfigs(t *testing.T) {
+	redisCfg := `{"is_redis_cluster": true,"serving_url":"online-storage.merlin.dev","redis_addresses":["10.1.1.10", "10.1.1.11"],"pool_size": 4,"max_retries": 1,"dial_timeout": "10s"}`
```
Comment (karzuo) [inline api/config/config_test.go:192]: same

File: api/pkg/transformer/executor/predictor.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+func NewMockModelPredictor(respBody types.JSONObject, respHeader map[string]string) *mockModelPredictor {
+	return &mockModelPredictor{
+		mockResponseBody:   respBody,
+		mockResponseHeader: respHeader,
+	}
+}
+
+var _ ModelPredictor = (*mockModelPredictor)(nil)
+
+func (mock *mockModelPredictor) ModelPrediction(ctx context.Context, requestBody []byte, requestHeader map[string]string) (respBody types.JSONObject, respHeaders map[string]string, err error) {
```
Comment (karzuo) [inline api/pkg/transformer/executor/predictor.go:33]: too long

File: api/pkg/transformer/executor/transformer.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+)
+
+type standardTransformer struct {
+	compiledPipeline *pipeline.CompiledPipeline
+	modelPredictor   ModelPredictor
+	executorConfig   transformerExecutorConfig
+	logger           *zap.Logger
+}
+
+// Transformer have predict function that process all the preprocess, model prediction and postproces
```
Comment (karzuo) [inline api/pkg/transformer/executor/transformer.go:23]: nitpick: missing 's' for postprocess

File: api/pkg/transformer/executor/transformer.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	traceEnabled         bool
+	transformerConfig    *spec.StandardTransformerConfig
+	featureTableMetadata []*spec.FeatureTableMetadata
+	feastOpts            feast.Options
+	logger               *zap.Logger
+	modelPredictor       ModelPredictor
+}
+
+// NewStandardTransformerWithConfig initialize standard transformer executor object
+func NewStandardTransformerWithConfig(ctx context.Context, transformerConfig *spec.StandardTransformerConfig, opts ...TransformerOptions) (Transformer, error) {
```
Comment (karzuo) [inline api/pkg/transformer/executor/transformer.go:38]: too long

File: api/pkg/transformer/pipeline/encoder_op.go [hunk 8]
```diff
@@ -44,6 +45,11 @@ func (e *EncoderOp) Execute(ctx context.Context, env *Environment) error {
 			return fmt.Errorf("encoder spec have unexpected type %T", encoderCfg)
 		}
 		env.SetSymbol(encoderSpec.Name, encoderImpl)
+		if e.OperationTracing != nil {
+			if err := e.AddInputOutput(nil, map[string]interface{}{encoderSpec.Name: "The result of this operation is on the transformer step that use this encoder"}); err != nil {
+				return err
```
Comment (karzuo) [inline api/pkg/transformer/pipeline/encoder_op.go:50]: too long

File: api/pkg/transformer/pipeline/feast_op.go [hunk 9]
```diff
@@ -8,14 +8,16 @@ import (
...
 )
 
 type FeastOp struct {
 	feastRetriever feast.FeatureRetriever
 	logger         *zap.Logger
+	*OperationTracing
 }
 
-func NewFeastOp(feastClients feast.Clients, feastOptions *feast.Options, entityExtractor *feast.EntityExtractor, featureTableSpecs []*spec.FeatureTable, logger *zap.Logger) Op {
+func NewFeastOp(feastClients feast.Clients, feastOptions *feast.Options, entityExtractor *feast.EntityExtractor, featureTableSpecs []*spec.FeatureTable, logger *zap.Logger, tracingEnabled bool) Op {
```
Comment (karzuo) [inline api/pkg/transformer/pipeline/feast_op.go:20]: line too long

File: api/pkg/transformer/pipeline/operation.go [hunk 10]
```diff
@@ -1,7 +1,121 @@
...
+
+// GetOperationTracingDetail retrieve all the tracing detail
+// it will flatten the result if the operation specs is in slice form
+func (ot *OperationTracing) GetOperationTracingDetail() ([]types.TracingDetail, error) {
+	refVal := reflect.ValueOf(ot.Specs)
+	if refVal.Kind() == reflect.Slice {
+		numOfSpecs := refVal.Len()
+
+		if len(ot.Input) != numOfSpecs {
+			return nil, fmt.Errorf("%s: number of inputs (%d) does not match with number of specs (%d)", ot.OpType, len(ot.Input), numOfSpecs)
```
Comment (karzuo) [inline api/pkg/transformer/pipeline/operation.go:85]: line too long

File: api/pkg/transformer/pipeline/operation.go [hunk 11]
```diff
@@ -1,7 +1,121 @@
...
+		}
+
+		if len(ot.Output) != numOfSpecs {
+			return nil, fmt.Errorf("%s: number of outputs (%d) does not match with number of specs (%d)", ot.OpType, len(ot.Output), numOfSpecs)
```
Comment (karzuo) [inline api/pkg/transformer/pipeline/operation.go:89]: line too long

File: api/service/transformer_simulation_service.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+		}),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed initializing standard transformer: %w", err)
+	}
+
+	return transformerExecutor, nil
+}
+
+func (ts *transformerService) simulate(ctx context.Context, transformer executor.Transformer, requestPayload types.JSONObject, requestHeaders map[string]string) (*types.PredictResponse, error) {
```
Comment (karzuo) [inline api/service/transformer_simulation_service.go:84]: line too long

**Comment by ariefrahmansyah** [comment]
> Generally LGTM except you tend to use very long lines which become hard to read/maintain.

For lines that are too long, you can enable word wrap in your IDE. For example https://code.visualstudio.com/docs/editor/codebasics#_how-do-i-turn-on-word-wrap

I have also formatted the code using gofumpt and husky and since they didn't modify it to the new line, I will keep it as is.

**Review by pradithya — APPROVED** [review]

**Pull request merged**